ostd/src/task/atomic_mode.rs

```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Atomic mode.
+//!
+//! Kernel is in atomic mode when it
+//! * runs in an IRQ handler,
+//! * keeps a [`crate::sync::SpinLock`],
+//! * explicitly disables preemption or IRQ.
+//!
+//! Kernel in atomic mode is supposed to run atomically (i.e., no preemption
+//! or blocking operation is allowed).
+
+#[cfg(debug_assertions)]
+use core::sync::atomic::Ordering;
+
+/// Indicates following blocking operations and panics if in atomic mode.
+pub fn might_break() {
+    #[cfg(debug_assertions)]
```

Comment:
I think the atomic mode should be enabled for release builds, too. This is because sometimes, doing sleep-like actions in atomic mode may have catastrophic consequences. Consider the following "bad" code.

```rust
let irq_line = IrqLine::acquire().unwrap();
irq_line.on_active(|_| {
    Task::yield_now();
});
```

Will such code cause UBs? Maybe. I don't know for sure. But what is certain is that our implementation of OSTD is not prepared for handling such safe-but-bad user code. Unlike Rust for Linux, Asterinas has a zero-tolerance policy for soundness holes. So we must take such safe-but-bad user code into consideration.

Reply:
OK I get it. The macros will be removed.

As to the code snippet you provided, I think it's exactly what the `non_block_*` things are all about. We often just don't want any operations triggering context-switch in atomic mode. However, there are some cases we are just OK with non-blocking operations. In these cases we are often not that sensitive to the turnaround time as we are in IRQ handlers so we allow operations like `yield` or `preempt` or IRQ but we dont't want a blocking operation which would likely cause an ABBA deadlock. Though, this may not have a downstream user in Asterinas so I removed it.

Reply:
The `non_block_*` things are a separate issue. Let's deal with them when we really need them.